radix-engine/src/blueprints/epoch_manager/validator.rs

```diff
@@ -456,6 +457,168 @@ impl ValidatorBlueprint {
+    /// call to [`finish_unlock_owner_stake_units()`].
+    /// As an optimization, this method also performs the same "finish unlocking" operation, and
+    /// returns the already-available stake units (if any).
+    pub fn start_unlock_owner_stake_units<Y>(
+        requested_stake_unit_amount: Decimal,
+        api: &mut Y,
+    ) -> Result<Bucket, RuntimeError>
+    where
+        Y: ClientApi<RuntimeError>,
+    {
+        // read the current epoch (needed for a drive-by "finish unlocking" of available withdrawals)
+        let epoch_manager_handle = api.actor_lock_field(
+            OBJECT_HANDLE_OUTER_OBJECT,
+            EpochManagerField::EpochManager.into(),
+            LockFlags::read_only(),
+        )?;
+        let epoch_manager: EpochManagerSubstate =
+            api.field_lock_read_typed(epoch_manager_handle)?;
+        let current_epoch = epoch_manager.epoch;
+        api.field_lock_release(epoch_manager_handle)?;
+
+        // read the configured unlock epochs delay
+        let config_handle = api.actor_lock_field(
+            OBJECT_HANDLE_OUTER_OBJECT,
+            EpochManagerField::Config.into(),
+            LockFlags::read_only(),
+        )?;
+        let config: EpochManagerConfigSubstate = api.field_lock_read_typed(config_handle)?;
+        let num_owner_stake_units_unlock_epochs = config.num_owner_stake_units_unlock_epochs;
+        api.field_lock_release(config_handle)?;
+
+        // begin the read+modify+write of the validator substate...
+        let handle = api.actor_lock_field(
+            OBJECT_HANDLE_SELF,
+            ValidatorField::Validator.into(),
+            LockFlags::MUTABLE,
+        )?;
+        let mut substate: ValidatorSubstate = api.field_lock_read_typed(handle)?;
+
+        // - drain the already-available withdrawals (as a drive-by)
+        let available_withdrawals = substate
+            .pending_owner_stake_unit_withdrawals
+            .range(..=current_epoch)
+            .map(|(epoch, available_amount)| (epoch.clone(), available_amount.clone()))
+            .collect::<Vec<_>>();
+        let mut total_already_available_amount = Decimal::zero();
+        for (epoch, available_amount) in available_withdrawals {
+            // no batch delete in a BTree
+            substate.pending_owner_stake_unit_withdrawals.remove(&epoch);
+            total_already_available_amount += available_amount;
+        }
+
+        // - insert the requested withdrawal as pending (if possible)
+        let currently_pending_withdrawals = substate.pending_owner_stake_unit_withdrawals.len();
+        if currently_pending_withdrawals >= OWNER_STAKE_UNITS_PENDING_WITHDRAWALS_LIMIT {
+            return Err(RuntimeError::ApplicationError(
+                ApplicationError::ValidatorError(
+                    ValidatorError::PendingOwnerStakeWithdrawalLimitReached,
+                ),
+            ));
+        }
+        substate
```

Comment:
From one point of view - yeah, it is arbitrary (we only have it because someone raised concern about iterating over thousands of items).

But at the same time, I think it's more elegant to return "sorry, you have too many pending withdrawals already" than to return "sorry, substate size limit reached".